### PR TITLE
Added info on gdt and volatile.

### PR DESCRIPTION
**Pull request by DeanoBurrito**
Let me know what you think.

### CODE REVIEW

File: C_Language_Info.md [hunk 1]
```diff
@@ -100,3 +100,27 @@ These are passed as integer args (registers, than the first if not enough space)
 To call an assembly function from C is pretty straight forward. If your assembly function takes arguments in the way described above, you can define a c prototype marked as `extern`, 
 and call it like any other function. In this case it's worth respecting the calling convention, and creating a new stack frame (`enter/leave` instructions).
 Any value placed in `eax/rax` will be returned from the c-function if it has a return type. It is otherwise ignored.
+
+## Use of `volatile`
+*__Note from the author of this section__: volatile is always a hot topic from c/c++ developers. The first page of any search involving it will usually have at least a few results like 'use of volatile considered harmful' thoroughly complaining about its existence.
+The next page is usually filled with an equal number of results of blog posts by people complaining about the previous people's complaing, and so on.
```
Comment (dreamos82) [inline C_Language_Info.md:106]: typo complaing -> complaining

File: GDT.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
+# The Global Descriptor Table
+
+## Overview
+
+The GDT is an x86(_64) structure that contains a series of descriptors. In a general sense, each of these descriptors tell the cpu about different things it should do. To refer to a GDT descriptor a selector is used, which is simply the byte offset from the beginning of the GDT where that descriptor starts.
+
+It's important to separate the idea of the bit-width of the cpu (16-bit, 32-bit, 64-bit) from the current mode (real mode, protected mode, long mode). Real mode is generally 16 bit, protected mode is generally 32 bit, and long mode is usually 64-bit, but this is not always the case. The GDT decides the bit-width (affecting how instructions are decoded, and how stack operations work for example), while CR0 and EFER affect the mode the cpu is in.
+
+Most descriptors are 8 bytes wide, usuallyt resulting in the selectors looking like the following:
```
Comment (dreamos82) [inline GDT.md:9]: typo: usuallyt -> usually

File: GDT.md [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+- ES: Extra selector, intended for use with string operations, no specific purpose.
+- FS: F selector, no specific purpose. Sys V ABI uses it for thread local storage.
+- GS: G selector, no specific purpose. Sys V ABI uses it for process local storage, commonly used for cpu-local storage in kernels due to `swapgs` instruction.
+
+Address types:
+- Logical address: addresses the programmer deals with.
+- Linear address: logical address after translation through segmentation (logical_address + selector_base).
+- Physical address: linear address translated through paging, maps to an actual memory location in RAM.
+
+It's worth noting if segmentation is ignored, logical and linear addresses are the same. <br>
```
Comment (dreamos82) [inline GDT.md:53]: is the <br> put there on purpose? (maybe it has a meaning and i don't know :D)

File: GDT.md [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+If paging is disabled, linear and physical addresses are the same.
+
+## Segmentation
+
+Segmentation is a mechanism for separating regions of memory into code and data, to help secure operating systems and hardware against potential security issues, and simpifies running multiple processes.
```
Comment (dreamos82) [inline GDT.md:58]: typo simpifies -> simplifies ?

File: GDT.md [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+How it works is pretty simple, each GDT descriptor defines a 'segment' in memory, using a base address and limit. 
+When a descriptor is loaded into the appropriate segment register, it creates a window into memory with the specified permissions. All memory outside of this segment has no persmissions (read, write, execute) unless specified by another segment.
```
Comment (dreamos82) [inline GDT.md:61]: typo persmissions -> permissions

File: GDT.md [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+Accessing memory inside a segment is done relative to it's base. Lets say you have a segment with a base of 0x1000,
+and some data in memory at address 0x1100.
+The data would be accessed at address 0x100 (assuming the segment is the active DS), as addressed are translated as `segment_base + offset`. In this case the segment base is 0x1000, and the offset is 0x100.
+
+Segments can also be explicitly referenced. To load something at offset 0x100 into the ES region, you can use `mov es:0x100, $rax`. This would perform the translation from logical address to linear address using ES instead of DS (the default for data), a common example is when an interrupt occurs while the cpu is in ring 3, it will switch to ring 0 and load the appropriate descriptors into the segment registers.
+
+### Segment Registers
+
+The various segment registers and their uses are outline below. There are some tricks to load a descriptor from the GDT into a segment register. They can't be mov'd into directly, so you'll need to use a scratch register to change their value. The cpu will also automatically reload segment registers on certain events (see the manual for these).
```
Comment (dreamos82) [inline GDT.md:73]: i'm not totally sure (you are the native speaker so maybe it's me being wrong) should be outlined below?

File: GDT.md [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+#example: load ds with the first descriptor
+mov $0x8, %ax       #any register will do, ax is used for the example here
+mov %ax, %ds
+
+#example: load ss with second descriptor
+mov $0x10, %ax
+mov %ax, %ss
+```
+
+Changing CS (code segment) is a little trickier, as it cant be written to directly, instead it requires a far jump:
```
Comment (dreamos82) [inline GDT.md:88]: can't?

File: C_Language_Info.md [hunk 8]
```diff
@@ -100,3 +100,27 @@ These are passed as integer args (registers, than the first if not enough space)
...
+
+Caching then adds more layers of complexity to this. Of course you can invalidate cache lines manually, however you'll pretty quickly find yourself fighting your compiler going this route. Best to stick to language constructs if you can.
+
+## The How
+`volatile` tells the compiler 'this variable is now observable behaviour', meaning it can still do clever tricks around this variable, but *at any point* the variable must exist in the exact state as described by the source code, in the expected location (ram, not a cpu register). Meaning that updates to the variable are written to memory immediately.
+
+This removes a lot of options for both the compiler and cpu in terms of what they can do with this data, so it's best used with care, and only when needed.
+
+## A Final Note
+`volatile` is not always the best tool for a job. For platform-specific things, like mmio, you'd likely want to make use of platform specific tools. For example, on x86 you can map memory as cache type UC (uncachable) for these regions, meaning reads and writes happen when you expect them to. This is entirely dependent on your use case though.
```
Comment (dreamos82) [inline C_Language_Info.md:126]: i'm not sure the example is referred to the platform speicific tools or to the volatile variables? 

I would suggest to add a real case scenario where a volatile variable is needed (probably the timer example can be a good candidate)

File: GDT.md [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+
+All the theory is great, but how to apply it? 
+A simple example is outline just below, for a simple 64-bit long mode setup you'd need
+- Selector 0x00: null
+- Selector 0x08: kernel code (64-bit, ring 0)
+- Selector 0x10: kernel data (64-bit)
+- Selector 0x18: user code (64-bit, ring 3)
+- Selector 0x20: user data (64-bit)
+
+To populate these entries, we'll use the following:
```
Comment (dreamos82) [inline GDT.md:117]: I would say before all the code why not place a table that explains all of the fields in a gdt entry? 

And if needed also another table that explains the bits of the address loaded with lgdt

**Review by dreamos82 — CHANGES_REQUESTED** [review]

File: GDT.md (hunk 6 above)
Comment (DeanoBurrito) [reply GDT.md:73]: you're correct on that one ;)

File: C_Language_Info.md (hunk 8 above)
Comment (DeanoBurrito) [reply C_Language_Info.md]: Added apic timer example, and clarified the meaning behind the framebuffer example.

File: GDT.md (hunk 9 above)
Comment (DeanoBurrito) [reply GDT.md:151]: Added a table and more details on the contents of the commonly used GDT descriptors.

**Review by dreamos82 — APPROVED** [review]

**Pull request merged**